src/adagrams.js

```diff
@@ -1,15 +1,191 @@
+    );
+    const randomLetter = letterArray[randomNumber];
+    if (letterPoolDeepCopy[randomLetter] > 0) {
+      hand.push(randomLetter);
+      letterPoolDeepCopy[randomLetter] -= 1;
+    }
+  }
+  return hand;
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
   // Implement this method for wave 2
+  const upperCaseInput = input.toUpperCase();
+  const inputLetterCount = {};
+  const lettersInHandCount = [];
+
+  for (const letter of lettersInHand) {
+    if (inputLetterCount[letter]) {
+      inputLetterCount[letter] += 1;
+    } else {
+      inputLetterCount[letter] = 1;
+    }
+  }
+
+  for (const letter of upperCaseInput) {
+    if (lettersInHandCount[letter]) {
+      lettersInHandCount[letter] += 1;
+    } else {
+      lettersInHandCount[letter] = 1;
+    }
+  }
+
+  for (const letter of input) {
+    if (inputLetterCount[letter] === lettersInHandCount[letter]) {
+      continue;
+    } else {
+      return false;
+    }
+  }
+  return true;
 };
 
 export const scoreWord = (word) => {
   // Implement this method for wave 3
+  if (word.length === 0) {
+    return 0;
+  }
+
+  const scoreDict = {
+    A: 1,
+    B: 3,
+    C: 3,
+    D: 2,
+    E: 1,
+    F: 4,
+    G: 2,
+    H: 4,
+    I: 1,
+    J: 8,
+    K: 5,
+    L: 1,
+    M: 3,
+    N: 1,
+    O: 1,
+    P: 3,
+    Q: 10,
+    R: 1,
+    S: 1,
+    T: 1,
+    U: 1,
+    V: 4,
+    W: 4,
+    X: 8,
+    Y: 4,
+    Z: 10,
+  };
+
+  let wordTotal = 0;
+  const upperCaseWord = word.toUpperCase();
+
+  console.log(word.length);
+  if (7 <= word.length && word.length <= 10) {
+    wordTotal += 8;
+  }
+
+  for (const letter of upperCaseWord) {
+    wordTotal += scoreDict[letter];
+  }
+  return wordTotal;
 };
 
 export const highestScoreFrom = (words) => {
   // Implement this method for wave 4
+  const wordArr = words
+    .map((word) => ({ word: word }))
+    .map((o) => {
+      o.score = scoreWord(o.word);
+      return o;
+    });
+
+  let highestScore = wordArr.reduce((maxScore, o) => {
+    if (o.score > maxScore) {
+      maxScore = o.score;
+    }
+    return maxScore;
+  }, -Infinity);
+
+  let highestScoringWords = [];
+
+  for (const o of wordArr) {
+    if (o.score === highestScore) {
+      highestScoringWords.push(o);
+    }
+  }
+  if (highestScoringWords.length === 1) {
+    return highestScoringWords[0];
+  } else if (highestScoringWords.length > 1) {
+    // sort highestScoringwords with shortest words first
+    // highestScoringWords.sort((a, b) => {
+    //   if (a.word.length < b.word.length) {
+    //     return -1;
+    //   }
+    //   if (a.word.length > b.word.length) {
+    //     return 1;
+    //   }
+    //   return 0;
+    // });
+    highestScoringWords.sort((a, b) => a.word.length - b.word.length);
+    // console.log(highestScoringWords);
+  }
+
+  for (const o of highestScoringWords) {
+    if (o.word.length >= 10) {
+      return o;
+    }
+  }
+
+  return highestScoringWords[0];
 };
```

Comment:
I understand the instinct to build a dictionary with all the scored words and then finding all the ones with the highest score and then making the comparisons separately, but this is one case where breaking everything down into its requisite pieces actually really muddles things and makes the algorithm less efficient. The pseudocode for a better approach is as follows:

      1. Create variables to hold the highest scoring word and the highest score.
      2. Loop through the word list and grab each word and its score.
      3. If its score is higher than the highest score, replace the highest scoring word and highest score.
      4. If there is a tie, do the comparisons for length in place within the loop itself.
      5. Once the loop is finished, you'll have the highest scoring word.